Fix MQTT retained message not being re-dispatched

OttoWinter opened the pull request:
## Description:

Fix MQTT retained message not being re-dispatched. See #12003 for a detailed explanation.

Fixes #12003

Unfortunately, I have very limited knowledge of the test framework Home Assistant uses and I haven't found a way to create tests for this PR - if somebody with more knowledge could help out that would be great.

## Checklist:

- [x] The code change is tested and works locally.

If the code communicates with devices, web services, or third-party tools:

- [x] Local tests with tox run successfully. **Your PR cannot be merged unless tests pass**.
- [x] Tests have been added to verify that the new code works.

OttoWinter committed:
Fix MQTT retained message not being re-dispatched

OttoWinter committed:
Fix tests

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -532,10 +533,21 @@ def async_subscribe(self, topic, qos):
         with (yield from self._paho_lock):
-            if topic in self.subscribed_topics:
+            # Don't re-subscribe to wanted subscription topics.
+            if topic in self.wanted_topics:
```
I'm using `wanted_topics` here, because it wouldn't make any sense to call the `paho-mqtt` subscribe method as it will soon be called anyway.

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -592,6 +605,13 @@ def _mqtt_on_subscribe(self, _mqttc, _userdata, mid, granted_qos):
     def _mqtt_on_message(self, _mqttc, _userdata, msg):
         """Message received callback."""
+        if msg.retain:
+            self.retained_messages[msg.topic] = (msg.payload, msg.qos)
```
I don't know too much about Home Assistant's style preferences concerning usage of tuples. Creating a namedtuple/struct/... here would probably be unnecessary since it's only really used in a single other place.

balloob commented:
Easy to reproduce with a test:

1. subscribe to a topic with listener A
2. mimic a retained message being send to that topic
3. assert listener A received message
4. subscribe to a topic with listener B
5. assert listener B received message

balloob reviewed (changes requested):
You wrote the one and only correct solution to this problem in #12003: don't re-use subscriptions but let paho-mqtt handle this.

If we handle it ourselves instead of paho-mqtt, we need to duplicate ALL retain logic that paho-mqtt already contains. That's not our responsibility.

OttoWinter commented:
@balloob You're right - at first when I tried the "clean" solution with the paho client, I think I saw a subscription not being re-sent and I didn't look into it any further. Though now I've tested this in more detail and the clean solution seems to work fine.

OttoWinter committed:
Use paho-mqtt for retained messages

OttoWinter committed:
Improve code style

balloob reviewed (changes requested):
How is unsubscription handled on the server? If we subscribe twice to the same topic, do we need to unsubscribe twice to not receive messages anymore? Or do we still need to handle that inside Home Assistant?

Right now if we subscribe a second time to the same topic, the retained message will also be sent to the existing subscriber. That is not correct.

OttoWinter commented:
@balloob Good catch! Looking at the code, this probably also means that, in the current Home Assistant, if we make two consecutive subscriptions to a topic and one of them is then unsubscribed, the remaining subscription is not re-established on a reconnect. I will look into this tomorrow.

balloob commented:
Correct, that is also a bug. Good thing that we don't unsubscribe a lot yet (will come when we start adding "unload component" functionality.

Can you add some test cases to test these situations?

OttoWinter committed:
Store list of subscribers

OttoWinter commented:
I've tried to characterize `paho-mqtt`'s (un)subscribe behavior and have come up with the following rules. But first, let's set up a notation to more easily describe the behavior:

* A message looks like this: `<topic> → <payload> [qos = 0]`;
  an ▶ arrow denotes a retained message.
  For example `'test/status' ▶ 'online'` is a retained message on the topic 'test/status' with payload 'online' and the default QoS 0.
* A subscribe call looks like this: `subscribe(<topic>, [qos=0])`
* An unsubscribe call looks like this: `unsubscribe(<topic>)`
* A (forced) reconnect is denoted by `reconnect()`
* Publishing a message is done by `publish(<message>)`
* A retained message `'test/status' ▶ 'online'` always exists.

1. `unsubscribe(<topic>)` deletes **all** subscriptions with the **exact** topic string.
    For example, if we do `subscribe('test/beer')`, `subscribe('test/beer')`, `subscribe('test/wine')`, `unsubscribe('test/beer')`, then we will receive `'test/wine' → '🍷'` but not `'test/beer' → '🍺'`
    However, when calling `subscribe('test/beer')`, `subscribe('test/+')`, `unsubscribe('test/beer')`, `'test/beer' → '🍺'` will still be received through `subscribe('test/+')`.
2. Every new message will always create **at most one** `on_message` callback call. An exception to this are re-sent retained messages on a `subscribe()`.
    So even though `subscribe('test/beer')` and `subscribe('test/+')`, we will only get one `on_message` for `'test/beer' → '🍺'`.
3. 1&2 also means that after all past retained messages have been re-sent on `subscribe()`, all subscriptions with the exact same topic behave as if they were one.

> Right now if we subscribe a second time to the same topic, the retained message will also be sent to the existing subscriber. That is not correct.

Is this really a problem? I mean, currently all retained messages are re-sent to all subscribers on `reconnect()`. Additionally, this would be very difficult to fix since we don't know whether a `on_message` call was triggered by a real new message or a `subscribe()` retained message. (If this is required, a fix might be to handle retained messages ourselves, though this is probably not the way to go)

My idea right now would have been to only allow unsubscribing through the return value of `async_subscribe` and modify `wanted_topics` to keep track of a **list** of active subscribers. Then we would only call `_mqttc.unsubscribe` when the list for a topic becomes empty.

test cases: Working on that right now. Unfortunately I know very little of python's unittest test framework so I first need to get a hang of it.

OttoWinter committed:
Fix lint error

OttoWinter committed:
Adhere to Home Assistant's logging standard

"Try to avoid brackets and additional quotes around the output to make it easier for users to parse the log."
 - https://home-assistant.io/developers/development_guidelines/

OttoWinter committed:
Add reconnect tests

OttoWinter committed:
Fix lint error

balloob commented:
I guess it's not a problem for retained messages to be sent twice. It should not happen too often.

OttoWinter commented:
That's good. I now also added some test cases that should cover the previous bug(s) - please tell me if I should change something there for example how I used the mocks.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -351,8 +361,8 @@ def async_setup(hass, config):
     # For cloudmqtt.com, secured connection, auto fill in certificate
-    if certificate is None and 19999 < port < 30000 and \
-       broker.endswith('.cloudmqtt.com'):
+    if (certificate is None and 19999 < port < 30000 and
```
Please don't include these "style" fixes in PRs. It makes it difficult to see what this PR actually adds.

If you want to do those, do it in a separate PR. Just know that generally, style fixes best case will make the code more readable, worst case introduce bugs that didn't exist before.

balloob replied:
For this PR it's fine, just don't do it in the future please.

OttoWinter replied:
Sure. I'll try to remember this.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -527,35 +537,52 @@ def stop():
+            return
+
+        subscriptions.remove(qos)
+        if not subscriptions:
```
Use a guard clause to save on indentation:

```python
if subscriptions:
    return

# Only unsubscribe …
```

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -246,8 +250,14 @@ def async_mqtt_topic_subscriber(dp_topic, dp_payload, dp_qos):
+    @callback
+    def async_remove_unsubscribe():
+        """Unsubscribe when removing signal listener."""
+        hass.async_run_job(hass.data[DATA_MQTT].async_unsubscribe(topic, qos))
```
It's messy that now we have 2 places where subscriptions live, that feels weird.

Probably out of scope, but it would almost make more sense to get rid of the dispatcher and move topic matching and calling listeners into our MQTT client. 

Something like this:

```python
sub_info = namedtuple('SubscriptionInfo', 'callback,qos,topic,encoding')

for sub in self._subscriptions.items():
    if not _match_topic(sub.topic, topic):
        return
    …
    self.hass.async_run_job(sub.callback, topic, payload, qos)
```

OttoWinter replied:
Yes, it's probably a good time to clean up the subscription model a bit. 🎨

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -527,35 +537,52 @@ def stop():
-            self._mqttc.unsubscribe, topic)
+        if not isinstance(topic, str):
+            raise HomeAssistantError("topic needs to be a string!")
+        if topic not in self.wanted_topics:
```
Silent failure (not raising when something is wrong) is usually a source of future bugs. The application clearly got into a wrong state or has a bug if we try to unsubscribe a subscription twice.

I know the old code did it too, but I think we should remove it. What do you think?

OttoWinter replied:
You're completely right - this has annoyed me with other components too so I don't know why I repeated this mistake. It's definitely better to fail fast than to have some unexpected behavior around - so I suppose raising an exception would be good to indicate a failed state.

Anyway, this could also be fixed by introducing a `Subscription` object as you mentioned. Working on that 🛠

balloob commented:
Your tests look great 👍

OttoWinter commented:
Thanks for the detailed review!

OttoWinter commented:
Sorry everybody! I accidentally merged instead of rebased...

OttoWinter committed:
Introduce Subscription

Tests still need to be updated

OttoWinter committed:
Use namedtuple for MQTT messages

... And fix issues

Accessing the config manually at runtime isn't ideal

OttoWinter committed:
Fix MQTT __init__.py tests

* Updated usage of Mocks
* Moved tests that were testing subscriptions out of the MQTTComponent test, because of how mock.patch was used
* Adjusted the remaining tests for the MQTT clients new behavior - e.g. self.progress was removed
* Updated the async_fire_mqtt_message helper

OttoWinter committed:
✅ Update MQTT tests

* Re-introduce the MQTT subscriptions through the dispatcher for tests - quite ugly though...  🚧
* Update fixtures to use our new MQTT mock 🎨

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -432,23 +424,29 @@ def async_publish_service(call):
-        self.wanted_topics = {}
-        self.subscribed_topics = {}
-        self.progress = {}
+        self.subscriptions = []
```
This list now holds all active subscriptions with the `Subscription` tuple defined before. When a new subscription arrives it's first added to this list and then the paho `subscribe` call is made.
On a reconnect, this is also used to re-create all subscriptions as before.

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -432,23 +424,29 @@ def async_publish_service(call):
+Subscription.__new__.__defaults__ = (0, 'utf-8')
+
+Message = namedtuple('Message', ['topic', 'payload', 'qos', 'retain'])
+Message.__new__.__defaults__ = (0, False)
```
While we're at creating `namedtuple`s, let's also create another struct to copy MQTT messages around.

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -351,17 +331,29 @@ def async_setup(hass, config):
+        will_message = Message(will_conf.get(ATTR_TOPIC),
+                               will_conf.get(ATTR_PAYLOAD),
+                               will_conf.get(ATTR_QOS),
+                               will_conf.get(ATTR_RETAIN))
```
The `Message` namedtuple also has the added benefit of not having to pass around configuration objects - please tell me if this is not wanted/should be in another PR.

balloob replied:
Fine to include in this PR.

balloob replied:
```python
will_message = Message(**will_conf)
```

OttoWinter replied:
Oh, that's very nice indeed!

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -39,7 +40,6 @@
 DATA_MQTT = 'mqtt'
 
 SERVICE_PUBLISH = 'publish'
-SIGNAL_MQTT_MESSAGE_RECEIVED = 'mqtt_message_received'
```
As all callbacks are now handled by the `MQTT` class, the dispatcher calls are no longer required (*except for tests)

balloob replied:
The tests should test the actual code, so let's make sure it is also not needed in tests.

OttoWinter replied:
Ok, I'll do that

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -526,36 +520,54 @@ def stop():
+        def async_remove():
+            """Remove subscription."""
+            if subscription not in self.subscriptions:
+                raise HomeAssistantError("Can't remove subscription twice")
```
As this line of code should only ever be executed when `async_remove` is called twice, I think it's good to really fail and not just put out an error log message.

OttoWinter commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -526,36 +520,54 @@ def stop():
-        self.wanted_topics.pop(topic, None)
-        result, mid = yield from self.hass.async_add_job(
-            self._mqttc.unsubscribe, topic)
+        with (yield from self._paho_lock):
```
This could potentially result in a race condition - so better safe than sorry and acquire the lock.

OttoWinter commented on tests/common.py:
```diff
@@ -212,13 +215,12 @@ def async_handle(self, intent):
-    dispatcher.async_dispatcher_send(
-        hass, mqtt.SIGNAL_MQTT_MESSAGE_RECEIVED, topic,
-        payload, qos)
+    dispatcher.async_dispatcher_send(hass, 'mqtt_message_received',
```
The return of the signal dispatcher ... Even though we don't use the dispatcher for MQTT during run time, we still need some way of firing MQTT events while running tests for MQTT-based components.

One way of doing so would be to mock the paho client and use the full MQTT client code - but that would (I think) go against the principle of *unit* testing. Therefore I propose that we test the internal MQTT callbacks using `mqtt/test_init.py` and use a mock with the dispatcher for all other component tests.

OttoWinter commented on tests/common.py:
```diff
@@ -291,14 +293,29 @@ def mock_http_component_app(hass, api_password=None):
+            if not mqtt._match_topic(topic, msg.topic):
+                return
+            sub = mqtt.Subscription(topic, msg_callback, 0, encoding)
+            mqtt._run_subscription_callback(hass, sub, msg)
```
(Using as much of the MQTT client code as possible here - this is still quite ugly though)

OttoWinter commented on tests/components/mqtt/test_discovery.py:
```diff
@@ -18,7 +18,7 @@ def test_subscribing_config_topic(hass, mqtt_mock):
     call_args = mqtt_mock.async_subscribe.mock_calls[0][1]
     assert call_args[0] == discovery_topic + '/#'
-    assert call_args[1] == 0
+    assert call_args[2] == 0
```
The method signature of `async_subscribe` in the internal `MQTT` client class was changed from `topic, qos` to `topic, msg_callback, qos, encoding`

OttoWinter commented on tests/conftest.py:
```diff
@@ -85,17 +85,9 @@ def aioclient_mock():
-        client.reset_mock()
-        return client
+    client = loop.run_until_complete(async_mock_mqtt_component(hass))()
+    client.reset_mock()
```
I'm not sure why `reset_mock()` is used here - leaving it there for now...

balloob replied:
It was used here to reset `client.mock_calls` to be empty. Since initializing the component calls the client, it came with added mocks that would never be tested upon.

OttoWinter replied:
Ah, that makes sense 👍

OttoWinter reviewed (commented):
The PR has now been updated to clean up the MQTT client code a bit and use a new `Subscription` object.

If you want I can also put the `Subscription` code into a new PR.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -632,6 +629,11 @@ def _mqtt_on_disconnect(self, _mqttc, _userdata, result_code):
 
+    def _grouped_subscriptions(self):
+        """Return a dict with all subscriptions grouped by their topics."""
+        iterator = groupby(self.subscriptions, lambda sub: sub.topic)
```
Make sure to sort `self.subscriptions` when using `groupby`

![image](https://user-images.githubusercontent.com/1444314/35945169-13fe0ac0-0c14-11e8-8178-4f5e59c08c6d.png)

I actually think that this function is quite overkill. For example, in `_mqtt_on_message`, it actually it more confusing what is going on compared to code that would just iterating over the list of subscriptions and matching each topic. I doubt it's also faster since you're now sorting the subscriptions, grouping them by topic and then converting them to a dictionary. Would that be faster than matching a few extra topics?

If you want to save on topic matching, you could consider converting `self.subscriptions` to be indexed by topic and as value have a list of subscriptions.

OttoWinter replied:
> Make sure to sort self.subscriptions when using groupby

That's good to know. Thanks! (The tests probably didn't catch that because it doesn't really break anything - just makes things a bit slower)

> I doubt it's also faster since you're now sorting the subscriptions, grouping them by topic and then converting them to a dictionary. Would that be faster than matching a few extra topics?

I based this on my own usage of MQTT - and I use the availability topic *everywhere*, so I have lots of subscriptions to the same topic - so in that case this would be quicker since the regex matching is probably (much) slower than constructing a dictionary (in `O(N log N)`). 

(*If* we're really out for performance, I could also implement a trie - but I mean this is python after all 🙃)

> it actually it more confusing what is going on compared to code

That's definitely true.

> you could consider converting self.subscriptions to be indexed by topic

I actually had that in before, but the problem with that was that you'd then have the topic in the `Subscription` object **and** the dictionary keys - could potentially cause problems.

 > code that would just [iterate] over the list of subscriptions and [match] each topic

Yes, I believe this is the best option and I'll change the code accordingly.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -663,6 +665,25 @@ def _match_topic(subscription, topic):
+def _run_subscription_callback(hass, subscription: Subscription,
+                               msg) -> None:
+    """Prepare message payload and run the subscription callback."""
+    if subscription.callback is None:
```
When can callback be `None` ?

OttoWinter replied:
I added this to allow subscriptions without callbacks - i.e. allow `async_subscribe` to be called to just cause a MQTT client subscription to be made without having to supply an empty callback. Right now this is only used in tests, so it's probably best to remove this "feature".

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -663,6 +665,25 @@ def _match_topic(subscription, topic):
+    if subscription.callback is None:
+        return
+
+    payload = msg.payload
```
Any reason this code is a standalone function?

balloob replied:
Nevermind, just continued reading.

balloob commented on homeassistant/components/mqtt/__init__.py:
```diff
@@ -571,50 +583,35 @@ def _mqtt_on_connect(self, _mqttc, _userdata, _flags, result_code):
-        if topic is None:
-            return
-        self.subscribed_topics.pop(topic, None)
+        _LOGGER.debug("Received message on %s: %s", msg.topic, msg.payload)
```
One thing that was nice about the `dispatcher` is that this code would be executed inside the event loop, which is a lot faster when calling `hass.async_add_job` (which happens inside `_run_subscription_callback`). There is quite some overhead to communicate between thread and event loop.

So, you could create a callback to be scheduled to handle matching etc. This would also help with the tests which would call that method directly.

```
    def _mqtt_on_message(self, _mqttc, _userdata, msg):
        self.hass.async_add_job(self._mqtt_handle_message(msg))

    @callback
    def _mqtt_handle_message(self, msg):
        …
```

OttoWinter committed:
📝 Update base code according to comments

OttoWinter committed:
🔨 Adjust MQTT test base

OttoWinter committed:
🔨 Update other MQTT tests

OttoWinter committed:
🍎 Fix carriage return in source files

Apparently test_mqtt_json.py and test_mqtt_template.py were written on Windows. In order to not mess up the diff, I'll just redo the carriage return.

OttoWinter committed:
🎨 Remove unused import

OttoWinter committed:
📝 Remove fire_mqtt_client_message

OttoWinter commented:
Your previous comments should now have been taken care of. To summarise the changes:

* `MQTT._grouped_subscriptions` has been removed and the `subscriptions` list is now simply iterated over without any grouping. Some of that code now resides in `_mqtt_on_connect` though, but the comments should explain what's going on well enough.
* `async_fire_mqtt_message` in the `tests/common` emulates MQTT messages by calling the `_mqtt_on_message` callback in the MQTT client and doesn't use the dispatcher any more.
* To make all tests work with above change, the MQTT class is now wrapped inside a `MagicMock`.
* Tests that previously used `mock_calls` on the MQTT client to find out whether a MQTT message was fired had to be updated (with an overly complicated regex...). The updated tests also have the added benefit of being a bit easier to read:

```python
# before
self.assertEqual(('alarm/command', 'ARM_HOME', 0, False),
                 self.mock_publish.mock_calls[-2][1])

# after
self.mock_publish.async_publish.assert_called_once_with(
    'alarm/command', 'ARM_HOME', 0, False)
```

(Side note: the mock mqtt *client* is stored as `self.mock_publish` in all test cases, that name seems a bit misleading to me.)

OttoWinter committed:
🐛 Fix using python 3.6 method

What's very interesting is that 3.4 didn't fail on travis...

OttoWinter committed:
🐛 Fix using assert directly

balloob commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -1395,53 +1395,60 @@ def test_state_changes_are_published_to_mqtt(self):
         self.hass.block_till_done()
-        self.assertEqual(('alarm/state', STATE_ALARM_DISARMED, 0, True),
-                         self.mock_publish.mock_calls[-2][1])
+        self.mock_publish.async_publish.assert_called_once_with(
```
These too are dangerous. Rename it to `assert_called_once_with_bla_bla` and the test will still pass.

balloob replied:
(means the old code was 👍)

Also, when doing a refactor, don't try to change the tests too. That makes it harder to validate that your code did not change anything.

OttoWinter replied:
I understand your concern. There were however reasons for this change. Let me explain.

> These too are dangerous. Rename it to `assert_called_once_with_bla_bla` and the test will still pass.

This is actually not entirely true. Are these asserts dangerous in principle (when used incorrectly)? Yes, very much so (*). Are they dangerous here? Not so much. For example, renaming the method to `assert_called_once_with_bla_bla` here will correctly raise an `AttributeError`.

<img width="824" alt="screen shot 2018-02-10 at 11 26 48" src="https://user-images.githubusercontent.com/6833237/36061401-ebdb8020-0e59-11e8-93d4-52847dd32d57.png">

This is because I'm **wrapping** the MQTT client object in a MagicMock and setting [**`spec_set`**](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock) as well. By doing this, "attempting to set or get an attribute on the mock that isn’t on the object passed as *spec_set* will raise an `AttributeError`."

(* Side Note: Starting with Python 3.5, `MagicMock`s have an argument called `unsafe` that when set to `False` (the default) causes an `AttributeError` to be raised when attempting a non-existant call on a attribute with the name starting with `assert` or `assret`)

> (means the old code was 👍)

When updating the base tests to use the actual MQTT client code and not the dispatcher, *something* changed and all `self.mock_publish.mock_calls[-2][1]` asserts were failing. Why did all of them suddenly fail? I didn't know at the time and it seemed very weird to me at the time since the *publish* code didn't really change. Why was the second last mock_call always accessed by the test code with `[-2]`? - I had no idea at all.

Of course, especially after the previous comment about code reviews being complicated by style fixes, I wanted to change as little as possible for the test code - after all, it would also mean quite a bit of work for me. Anyway, so I spend quite a while trying to debug the problem and (only partially successfully) spinning up a debugger (I'm no Python developer, to say the least) - without finding a solution using mock_calls. And after a while I found the "potential" fix of just using `assert_called_once_with`, which to me also seemed to have the added benefit of being much easier to read - especially so since I personally had struggled quite a bit understanding what `[-2]` was trying to indicate.

Frankly, I have to say that at the time, I didn't know about the potential pitfalls of using `assert_called_*` and setting `spec_set` was mostly *accidental* and of the motto: "it can't hurt".

I know now how I could go about "fixing" the tests without changing too much code - simply changing `-2` to `-1` in `self.mock_publish.mock_calls[-2][1]` *seems* to work. I do, however, *not* know *why* the `-2` was there in the first place - that would definitely be good to know before going about fixing it that way.

> That makes it harder to validate that your code did not change anything.

I completely understand. I hope you understand that at the time and after unsuccessfully trying to fix the tests, it seemed like the best option to me.

Now, I can, if you want to, change the tests to use `[-1]` everywhere or if you know why `-2` was used everywhere, I could also try to fix it that way.

balloob replied:
Ok, I didn't know about the `spec_set`, in that case it's better as it's more clear what is happening.

The reason things like -2 etc were used was because some other code was also calling something on the MQTT mock.

BTW, instead of going all fancy with the debugger, just `print(self.mock_publish.mock_calls)` will show a lot :)

OttoWinter replied:
Yeah I also stumbled upon that by accident, but it's quite useful. And not using `print(...)` at first was kind of dumb of me but hey you learn something new every day 😊

balloob reviewed (approved):
This is ok to merge once the merge conflicts have been resolved

OttoWinter commented:
Done 🎉
